src/translate.c

```diff
@@ -82,21 +82,39 @@ static bool list_any_known_encoding(SEXP x, R_len_t size);
 static bool df_any_known_encoding(SEXP x, R_len_t size);
 
 static bool obj_any_known_encoding(SEXP x, R_len_t size) {
-  switch (vec_proxy_typeof(x)) {
-  case vctrs_type_character: return chr_any_known_encoding(x, size);
-  case vctrs_type_list: return list_any_known_encoding(x, size);
-  case vctrs_type_dataframe: return df_any_known_encoding(x, size);
-  default: return false;
+  switch (TYPEOF(x)) {
+  case STRSXP: {
+    return chr_any_known_encoding(x, size);
+  }
+  case VECSXP: {
+    if (is_data_frame(x)) {
```

Comment:
Maybe add a comment that dfs and lists are treated differently for performance, because we can make size assumptions on df columns? IIUC.